### PR TITLE
(explicitly) find the executable before running run_cmd

### LINKED ISSUE #3035: git.bat not picked up on dbt deps
ref. https://github.com/fishtown-analytics/dbt/pull/1110

```
> dbt deps
Encountered an error:
Could not find command, ensure it is in the user's PATH: "git"
```

I have reasons to wrap my executables. On Windows I wrap my executables in `.bat` scripts (to get a similar effect to `.sh`). But then[ `Popen( ['git'] )`](https://github.com/fishtown-analytics/dbt/blob/42a85ac39f34b058678fd0c03ff8e8d2835d2808/core/dbt/clients/system.py#L419) doesn't work. `Popen( ['git.bat'])` works. 

So the platform-independent way to get `git` to work is to use  `shutil.which` to get the full path to the exe. `Popen([ which('git') ], )` works.

### PR DESCRIPTION
**Pull request by majidaldo**
resolves #3035

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
   ran test_system_client.py
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by jtcohen6** [comment]
Thanks for this @majidaldo!

- Can you confirm that this change resolves #3035 for you when running `dbt deps` locally?
- It looks like there are two failing unit tests: [`test__executable_does_not_exist`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L76) and [`test__not_exe`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L86). My guess here is that these tests are expecting an [`ExecutableError`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/core/dbt/exceptions.py#L392) instead of a [`CommandError`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/core/dbt/exceptions.py#L379).
- Changelog: By "dbt next," we just mean the next (not-yet-released) version of dbt, which right now is `dbt 0.20.0 (Release TBD)`. Could you add an entry beneath "Fixes" or "Under the hood" (whichever feels more appropriate) describing the change, with links to the issue and this PR in parentheses, and also add yourself to the list of Contributors?

@kwigley I'm going to tag you as a reviewer for a second (better-informed) pair of eyes. The proposed change is quite narrow, and I don't think it raises any security implications.

File: core/dbt/clients/system.py [hunk 1]
```diff
@@ -407,6 +407,11 @@ def run_cmd(
     logger.debug('Executing "{}"'.format(' '.join(cmd)))
     if len(cmd) == 0:
         raise dbt.exceptions.CommandError(cwd, cmd)
+    exe_pth = shutil.which(cmd[0])
+    if not exe_pth:
+        # executable not found
+        raise dbt.exceptions.CommandError(cwd, cmd)
```
Comment (kwigley) [inline core/dbt/clients/system.py:413]: I think we want to raise an `ExecutableError` here instead (like @jtcohen6 mentioned) there are some places where we present relevant error messages for an `ExecutableError` with git commands

**Review by kwigley — APPROVED** [review]
👍 glad this was an easy fix, just asking to update the relevant exception raised but looks good otherwise

**Review by kwigley — CHANGES_REQUESTED** [review]
whoops sorry! I meant to request changes

**Comment by majidaldo** [comment]
> * Can you confirm that this change resolves #3035 for you when running `dbt deps` locally?


Yes. I created a git.bat with `"C:\path\to\git.exe" %*` and put it in my current directory and `dept deps` worked.  When I change it to `"C:\path\to\notgit.exe" %*` it expectedly errors.

**Comment by jtcohen6** [comment]
@majidaldo Thanks for updating! Could you add yourself to the list of Contributors in the Changelog as well?

In the CI, it looks like there's a new minor flake8 (style) error:
```
core/dbt/clients/system.py:412:80: E501 line too long (93 > 79 characters)
```

It also looks like these unit tests ([`test__executable_does_not_exist`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L76) and [`test__not_exe`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L86)) are still failing. In particular, even though it's now correctly raising an `ExecutableError`, the expected words `'path'` and `'permissions'` are not present in the returned error message.

In the CI runs, I'm seeing error messages like:
```
executable for /tmp/tmpi8g4qwbh/does_not_exist not found.: "/tmp/tmpi8g4qwbh/does_not_exist"
executable for /tmp/tmp1xj494qp/empty_file not found.: "/tmp/tmp1xj494qp/empty_file"
```
Versus the expected error messages on POSIX:
```
Could not find command, ensure it is in the user's PATH
User does not have permissions for this command
```
Or Windows:
```
Could not find command, ensure it is in the user's PATH and that the user has permissions to run it
```

The latter error messages do seem a bit more specific and helpful. I'm wondering if the `if not exe_path` case, instead of adding its own custom error handling, should either shell out to `_interpret_oserror` or pass (?) and allow a more specific error to happen further down the code path. What do you think?

**Comment by majidaldo** [comment]
> @majidaldo Thanks for updating! Could you add yourself to the list of Contributors in the Changelog as well?
> 
> In the CI, it looks like there's a new minor flake8 (style) error:
> 
> ```
> core/dbt/clients/system.py:412:80: E501 line too long (93 > 79 characters)
> ```
> 
> It also looks like these unit tests ([`test__executable_does_not_exist`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L76) and [`test__not_exe`](https://github.com/fishtown-analytics/dbt/blob/344a14416d22f0cfbeb56b9904092c8a4f38b1fc/test/unit/test_system_client.py#L86)) are still failing. In particular, even though it's now correctly raising an `ExecutableError`, the expected words `'path'` and `'permissions'` are not present in the returned error message.
> 
> In the CI runs, I'm seeing error messages like:
> 
> ```
> executable for /tmp/tmpi8g4qwbh/does_not_exist not found.: "/tmp/tmpi8g4qwbh/does_not_exist"
> executable for /tmp/tmp1xj494qp/empty_file not found.: "/tmp/tmp1xj494qp/empty_file"
> ```
> 
> Versus the expected error messages on POSIX:
> 
> ```
> Could not find command, ensure it is in the user's PATH
> User does not have permissions for this command
> ```
> 
> Or Windows:
> 
> ```
> Could not find command, ensure it is in the user's PATH and that the user has permissions to run it
> ```
> 
> The latter error messages do seem a bit more specific and helpful. I'm wondering if the `if not exe_path` case, instead of adding its own custom error handling, should either shell out to `_interpret_oserror` or pass (?) and allow a more specific error to happen further down the code path. What do you think?

I'm afraid my 'scope' is to just offer my help for a small fix. I'd really need to dig into the design of dbt to give you a full answer.

**Comment by jtcohen6** [comment]
> I'm afraid my 'scope' is to just offer my help for a small fix. I'd really need to dig into the design of dbt to give you a full answer.

@majidaldo Got it — would you like us to take the code from here? I'm hopeful there's just a little bit of cleanup left.

**Comment by majidaldo** [comment]
> > I'm afraid my 'scope' is to just offer my help for a small fix. I'd really need to dig into the design of dbt to give you a full answer.
> 
> @majidaldo Got it — would you like us to take the code from here? I'm hopeful there's just a little bit of cleanup left.

Please. I hope I pointed out the usefulness of using `which` which (hah) helps with diagnosing problems on both linux and windows.

Thanks for the great work!

**Comment by jtcohen6** [comment]
@majidaldo Could you take #3299 for a spin? I think I managed to accomplish the same functional change, while preserving the original error messages. I'd like to confirm with you first, however, before merging!

**Comment by majidaldo** [comment]
> @majidaldo Could you take #3299 for a spin? I think I managed to accomplish the same functional change, while preserving the original error messages. I'd like to confirm with you first, however, before merging!

Yup! `dbt deps`:
* w/o git gives the expected message
* w/ git.exe works
* w/ git.bat works
 

thanks!

**Pull request closed without merging**